### ISSUE TITLE
[firebase_database][firebase_animated_list] 'assert', 'throw', 'throwIf' statements do not exit the FirebaseAnimatedList widget.

### ISSUE THREAD

**Issue by GrahamDi** [FirebaseExtended/flutterfire#1897]
**Describe the bug**
I am attempting to debug the chat messaging function of my app, which makes use of FirebaseAnimatedList (with sort option). The reason for needing to debug is that I am experiencing the following error message, on occasion, which I believe is being caused by a corrupt (incomplete) error message returned from the db in the snapshot:
> E/flutter ( 5081): [ERROR:flutter/flow/layers/transform_layer.cc(22)] TransformLayer is constructed with an invalid matrix.

When this error occurs, the app appears to freeze although I am sure it is looping internally (Flutter engine maybe). Obviously, I need to fix this before going live.

My normal debugging approach is to use the VSCode debug facility and to include `print` and `assert `statements at strategic points in the code. However, when I use `assert `or even `throw` in the  `itemBuilder:` section of the `FirebaseAnimatedList` widget defintion, the app does not terminate but instead shows the same error on the console as above. I can see that an `assert ` or `throw `is being triggered by following the progress in the debugger and therefore it is now the `assert ` or `throw `and the subsequent break out of that cycle of the loop that is producing the same error message.  Problem is that `itemBuilder` continues to build and go round the loop again...which then either freezes the app or loops internally (Flutter engine). The print statement works fine, of course. Hope this is clear.
Surely, the `assert `or `throw `should exit the `itemBuilder` completely? It seems to do so in a `ListView.builder` `itemBuilder:` section with no issues.

**To Reproduce**
Here is a code snippet from my app. None of the `assert `or `throw `statements terminate the `itemBuilder `loop although they are triggering when I manipulate the message data to meet their logic.
``` dart
FirebaseAnimatedList(
  query: <my messages firebase reference>,
  padding: const EdgeInsets.all(8.0),
  reverse: true,
  sort: (a, b) => (b.value["srtTs"] as int)
          .compareTo(a.value["srtTs"] as int),
  itemBuilder: (_, DataSnapshot messageSnapshot,
      Animation<double> animation, int index) {
    _message = null;
    assert(messageSnapshot != null);
    _message = Message.fromSnapshot(messageSnapshot);
    assert(_message.senderIdInPositionOne != null);
    assert(_message.sortTimestamp != null);
    assert(_message.timestamp != null);
	
    print("_message: $_message");
	
    throwIf(_message.senderIdInPositionOne == null, "thrown");
    throw "Test throw in itemBuilder";

    return ChatMessageListItem(
      message: _message,
      animation: animation,
    );
  }
);
```

**Expected behavior**
I would expect the `assert `or `throw `to exit the `itemBuilder` loop and the FirebaseAnimiatedList widget completely. I could then catch and process the error in the `try `/ `catch `construct that surrounds this code in my app. I have to be able to terminate the build loop and throw a proper error in the app when I detect corrupt messages.

**Comment by iapicca** [comment]
Hi @GrahamDi 
Please see https://flutter.dev/community for resources and asking questions like this,
you may also get some help if you post it on Stack Overflow.
Closing, as this isn't an issue with FlutterFire itself,
if you disagree please write in the comments and I will reopen it
Thank you

**Comment by GrahamDi** [comment]
@iapicca Thanks for your response and I will do as you suggest. However, I don't see why this is not an issue with 'FlutterFire itself'.

FirebaseAnimatedList is a class within the firebase_animated_list library, which is itself part of the firebase_database library, which is listed on your FlutterFire read me page as one of the 'Available FlutterFire plugins'  (https://github.com/FirebaseExtended/flutterfire/blob/master/README.md)

Please explain to me how it is not a FlutterFire issue, therefore?

**Comment by iapicca** [comment]
Hi @GrahamDi 
can you please provide your `flutter doctor -v` , 
your `flutter run --verbose` 
and your `pubspec.yaml` 
or if possible a reproducible minimal code sample.
Thank you

**Comment by GrahamDi** [comment]
@iapicca Thanks. I will get something back to you within the next 24 hours.

**Comment by GrahamDi** [comment]
@iapicca 

I decided to do a bit more investigation by trying ```assert(true==false);``` in the ```itemBuilder:``` property of a ```ListView.builder``` surrounded by a try / catch structure and found the same issue. I had assumed that the assert or throw would immediately break out of the loop to the catch block. This is not the case. 

In order for asserts and throws to work within the ```itemBuilder:``` property, the try / catch has to be within ```itemBuilder:``` as well. So I have learnt something and your are quite right, this is not a FlutterFire issue :-)  Apologies. 

Closing this issue now.
